### DISCUSSION TITLE
How would one go about caching with Redis?

### DISCUSSION THREAD

**Discussion by gevera** [noam-honig/radweb#222] [General]
I would like to make Remult even better by adding to it some caching like Redis. What would be the best approach for that?

**Comment by YeudaBy** [comment]
Yes! it would be great to add some kv support!

**Comment by noam-honig** [comment]
Can you elaborate more on this - 
what would you like to cache? 
where would you like to cache it?
what kv support are you looking for?

**Comment by YeudaBy** [comment]
For now, i've been looking for support to vercel kv, but from what I've
gotten to know your library, I'm sure you'll do it in a way I can't
imagine...

Let's say, I could set the tv service at the remult config, and the best
think, in my mind, is to use it under the hood.

For exaple:
UserRepo.find({id: 2})
-> kv - "user-2", or from db..

בתאריך יום ה׳, 29 באוג׳ 2024, 09:58, מאת Noam Honig ‏<
***@***.***>:

**Comment by noam-honig** [comment]
Ok - and your goal would be to "save" calls to the db in this case?

**Comment by YeudaBy** [comment]
Save connections and get faster results

בתאריך יום ה׳, 29 באוג׳ 2024, 12:23, מאת Noam Honig ‏<
***@***.***>:

**Comment by noam-honig** [comment]
I see - this can be easily implemented by a custom dataProvider that delegates the wok if the value is not cached.

You can use the findOptions as sort of key - or the sql statement itself as sort of ket.

This can be done in userland, and we may add it in the future - if you run into a production issue that requires it, let me know and Il'l be happy to help

**Comment by jycouet** [comment]
Yeah, userland is a great place to start. _To try things out and see how good it is 👍_

For more context, you have today some performance issues? Quantity of user? Complexe relations? What is your stack?
Because in the end, it's also adding a new component to the stack, adding some questions around caching invalidation & co.

**Comment by noam-honig** [comment]
Hi Guys,

Just had a long session with a friend, that had queries that take seconds - and are re-executed multiple times by multiple users (and live query)

I've created a simple key value cache for it that keeps a cache for 60 seconds, for a specific entity, and invalidate the cache in case of update or delete to that entity - or related entities (this is based on a complex view)

For the key of the cache I used the query string + the body (he is using an old version of remult so I couldn't release updates)
TODO  later:
I think that the cache key, should be the where json + limit + orderby + page and in sql entities, also the sql expressions.

Here's the code that I've used there - let me know if that's close to what you had in mind - and if you would like to persue this further:

https://gist.github.com/noam-honig/9195316391784b9f34f09d2035b473e2
